test/python/circuit/library/test_qaoa_ansatz.py

```diff
@@ -0,0 +1,85 @@
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""Test QAOA ansatz from the library."""
+from qiskit.circuit.quantumcircuit import QuantumCircuit
+from qiskit.circuit.library.n_local.qaoa_ansatz import QAOAAnsatz
+from qiskit.circuit.library import HGate, RXGate, YGate, RYGate, RZGate
+from qiskit.opflow import I, Y, Z
+from qiskit.test import QiskitTestCase
+
+
+class TestQAOAAnsatz(QiskitTestCase):
+    """Test QAOAAnsatz."""
+    def test_default_qaoa(self):
+        """Test construction of the default circuit."""
+        circuit = QAOAAnsatz(I, 1)
+
+        parameters = circuit.parameters
+
+        self.assertEqual(1, len(parameters))
+        self.assertIsInstance(circuit.data[0][0], HGate)
+        self.assertIsInstance(circuit.data[1][0], RXGate)
+
+    def test_custom_initial_state(self):
+        """Test circuit with a custom initial state."""
+        initial_state = QuantumCircuit(1)
+        initial_state.y(0)
+        circuit = QAOAAnsatz(initial_state=initial_state, cost_operator=I, reps=1)
+
+        parameters = circuit.parameters
+        self.assertEqual(1, len(parameters))
+        self.assertIsInstance(circuit.data[0][0], YGate)
+        self.assertIsInstance(circuit.data[1][0], RXGate)
+
+    def test_custom_circuit_mixer(self):
+        """Test circuit with a custom mixer as a circuit"""
+        mixer = QuantumCircuit(1)
+        mixer.ry(1, 0)
+        circuit = QAOAAnsatz(cost_operator=I, reps=1, mixer_operator=mixer)
+
+        parameters = circuit.parameters
+        self.assertEqual(0, len(parameters))
+        self.assertIsInstance(circuit.data[0][0], HGate)
+        self.assertIsInstance(circuit.data[1][0], RYGate)
+
+    def test_custom_operator_mixer(self):
+        """Test circuit with a custom mixer as an operator."""
+        mixer = Y
+        circuit = QAOAAnsatz(cost_operator=I, reps=1, mixer_operator=mixer)
+
+        parameters = circuit.parameters
+        self.assertEqual(1, len(parameters))
+        self.assertIsInstance(circuit.data[0][0], HGate)
+        self.assertIsInstance(circuit.data[1][0], RYGate)
+
+    def test_all_custom_parameters(self):
+        """Test circuit with all custom parameters."""
+        initial_state = QuantumCircuit(1)
+        initial_state.y(0)
+        mixer = Z
+
+        circuit = QAOAAnsatz(cost_operator=I, reps=2, initial_state=initial_state,
+                             mixer_operator=mixer)
+
+        parameters = circuit.parameters
+        self.assertEqual(2, len(parameters))
+        self.assertIsInstance(circuit.data[0][0], YGate)
+        self.assertIsInstance(circuit.data[1][0], RZGate)
+        self.assertIsInstance(circuit.data[2][0], RZGate)
+
```

Comment:
Could you add a test where you change the attributes and make sure that the circuit is correctly rebuilt?

Reply:
What kind of attributes? I don't get what is missing.

Reply:
If you initialize the `QAOAAnsatz` with some parameters you can go and use that circuit, say
```python
ansatz = QAOAAnsatz(Z ^ I)  # circuit with 2 qubits
```
But since it is a `BlueprintCircuit` you can now change some attributes and the circuit should rebuild itself, e.g.
```
ansatz.cost_operator = Z  # now it only has 1 qubit
ansatz.reps = 5  #  and now 5 repetitions
```
I think this rebuilding is not yet covered, or is it?

Reply:
The circuit is not built across these three lines. Ok, let me see what can be added here.

Reply:
A new test is ready as discussed.